File: ios/Showcase/Targets/AppUIKit/Sources/wallet/HandleInteractionView.swift
import SpruceIDMobileSdkRs
import SwiftUI

struct HandleInteraction: Hashable {
    var url: String
    var credentialPackId: String?
}

enum DiscoveryError: Error {
    case invalidInteractionURL(reason: String)
}

struct HandleInteractionView: View {
    @Binding var path: NavigationPath
    let url: String
    var credentialPackId: String?

    @State var sheetOpen: Bool = false
    @State var err: String?
    @State var protocolErrorTitle: String?
    @State var protocolErrorDetails: String?
    @State var protocolSelected: Bool = false
    enum ExchangeError: Error {
        case badStatus(Int)
        case noProtocolsOrFallback
    }

    struct ProtocolsResponse: Decodable {
        let protocols: [String: String]
    }

    @State var protocols: [String: String] = [:]

    func loadProtocols() async {
        do {
            var discoveryUrl = url
            let options: String.CompareOptions = [.anchored, .caseInsensitive]

            // Remove "interaction:" prefix if present
            if let range = discoveryUrl.range(
                of: "interaction:",
                options: options
            ) {
                discoveryUrl = String(discoveryUrl[range.upperBound...])
            }
            protocols = try await discoverProtocols(
                interactionUrl: discoveryUrl
            )
            sheetOpen = true
        } catch {
            err =
                "Couldn't discover protocols from QR Code payload. Error: \(error)"
        }
    }
    func onBack() {
        while !path.isEmpty {
            path.removeLast()
        }
    }
    var body: some View {
        ZStack {
            if let protocolErrorTitle, let protocolErrorDetails {
                ErrorView(
                    errorTitle: protocolErrorTitle,
                    errorDetails: protocolErrorDetails,
                    onClose: onBack
                )
            } else if err != nil {
                ErrorView(
                    errorTitle: "Error discovering protocols",
                    errorDetails: err!,
                    onClose: onBack
                )
            } else {
                LoadingView(loadingText: "Discovering protocols...")
            }
        }.task {
            await loadProtocols()
        }.sheet(
            isPresented: $sheetOpen,
            onDismiss: {
                if !protocolSelected {
                    onBack()
                }
            }
        ) {
            ProtocolSelectorBottomSheet(
                protocols: $protocols,
                sheetOpen: $sheetOpen,
                protocolSelected: $protocolSelected,
                path: $path,
                credentialPackId: credentialPackId,
                onError: { title, details in
                    sheetOpen = false
                    protocolErrorTitle = title
                    protocolErrorDetails = details
                }
            )
            .presentationBackgroundInteraction(.automatic)
        }
        .navigationBarBackButtonHidden(true)
    }
}

File: android/Showcase/src/main/java/com/spruceid/mobilesdkexample/wallet/HandleInteractionView.kt
package com.spruceid.mobilesdkexample.wallet

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.navigation.NavController
import com.spruceid.mobile.sdk.rs.discoverProtocols
import com.spruceid.mobilesdkexample.ErrorView
import com.spruceid.mobilesdkexample.LoadingView
import com.spruceid.mobilesdkexample.navigation.Screen


@Composable
fun HandleInteractionView(
    navController: NavController,
    url: String,
    credentialPackId: String?
) {
    var protocols by remember { mutableStateOf<Map<String, String>>(emptyMap()) }
    var sheetOpen by remember { mutableStateOf(false) }
    var errorTitle by remember { mutableStateOf<String?>(null) }
    var errorDescription by remember { mutableStateOf<String?>(null) }

    suspend fun loadProtocols() {
        try {
            var discoveryUrl = url
            // Remove `interaction:` prefix if present
            if (discoveryUrl.startsWith("interaction:", ignoreCase = true)) {
                discoveryUrl = discoveryUrl.drop("interaction:".length)
            }
            protocols = discoverProtocols(discoveryUrl)
            sheetOpen = true

        } catch (e: Exception) {
            errorTitle = "Error discovering protocols"
            errorDescription = "Couldn't discover protocols from QR Code payload ${url}. Error: ${e.message}"
        }
    }

    LaunchedEffect(Unit) {
        loadProtocols()
    }

    fun onBack() {
        navController.navigate(Screen.HomeScreen.route) { popUpTo(0) }
    }


    if (sheetOpen) {
        ProtocolSelectorBottomSheet( { onBack() }, protocols=protocols, navController=navController, credentialPackId=credentialPackId)
    } else if (errorTitle != null && errorDescription != null) {
        ErrorView(
            errorTitle = errorTitle!!,
            errorDetails = errorDescription!!,
            onClose = { onBack() }
        )
    } else {
        LoadingView(loadingText = "Discovering protocols...")
    }
}





